geo/src/main/java/com/tomtom/speedtools/geometry/GeoPolyLine.java

```diff
@@ -50,26 +50,38 @@ public GeoPolyLine(@Nonnull final List<GeoPoint> points) {
         assert points != null;
         assert points.size() >= 2;
 
-        // Find the first elevation.
-        double elevationMeters = Double.NaN;
-        for (final GeoPoint point : points) {
-            if (point.getElevationMeters() != null) {
-                elevationMeters = point.getElevationMetersOrNaN();
-                break;
-            }
-        }
+        // Check the need to correct the elevation in a list of points.
+        final List<GeoPoint> elevatedPoints;
+        final boolean noneElevated = points.stream().allMatch(x -> (x.getElevationMeters() != null));
+        final boolean allElevated = points.stream().allMatch(x -> (x.getElevationMeters() == null));
 
-        // Create a new list of points, append first elevation if needed.
-        final List<GeoPoint> elevatedPoints = new ArrayList<>(points.size());
-        for (final GeoPoint point : points) {
+        if (allElevated || noneElevated) {
+
+            // Don't need to create a new list.
+            elevatedPoints = points;
+        } else {
 
-            // Use new elevation if available.
-            if (point.getElevationMeters() != null) {
-                elevationMeters = point.getElevationMeters();
+            // Find the first elevation.
+            double elevationMeters = Double.NaN;
+            for (final GeoPoint point : points) {
+                if (point.getElevationMeters() != null) {
+                    elevationMeters = point.getElevationMetersOrNaN();
+                    break;
+                }
             }
 
-            // Add point, with elevation added (might be its own elevation.
-            elevatedPoints.add(point.withElevationMeters(elevationMeters));
+            // Create a new list of points, append first elevation if needed.
+            elevatedPoints = new ArrayList<>(points.size());
+            for (final GeoPoint point : points) {
+
+                // Use new elevation if available.
+                if (point.getElevationMeters() != null) {
+                    elevationMeters = point.getElevationMeters();
+                }
+
+                // Add point, with elevation added (might be its own elevation.
+                elevatedPoints.add(point.withElevationMeters(elevationMeters));
+            }
```

Comment:
I do not really like that the order of the points determines which elevation is used. But I consider it as a very minor issue.